### PR TITLE
[PS-1878] Feature/selfhost anonaddy

### PR DESCRIPTION
**Pull request by Djcd**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Add custom url for support self hosted anonaddy instances in the username generator.

## Code changes
[apps/browser/src/popup/generator/generator.component.html](https://github.com/bitwarden/clients/compare/master...Djcd:clients:feature/selfhost-anonaddy?expand=1#diff-f9bc551eade4aa52ac1f81fc943ff55b1801b445535f0fa0a02eeb5ce26fbf52): Add field for custom url
[apps/desktop/src/app/vault/generator.component.html](https://github.com/bitwarden/clients/compare/master...Djcd:clients:feature/selfhost-anonaddy?expand=1#diff-6db850571b90ac2c8623542f33d393cbb5ff45cd409af107c154780bd6111abd): Add field for custom url
[libs/common/src/emailForwarders/anonAddyForwarder.ts](https://github.com/bitwarden/clients/compare/master...Djcd:clients:feature/selfhost-anonaddy?expand=1#diff-5c44f39060767d54c4c9e239ed075441c7e822e7164bc1c3d36dbb6a03700215): Add check for empty url and modify url variable for api request
[libs/common/src/emailForwarders/forwarderOptions.ts](https://github.com/bitwarden/clients/compare/master...Djcd:clients:feature/selfhost-anonaddy?expand=1#diff-2ebeb91cc4277a9546946dc0c44bf2d4851433ea01e58022045bb0b5349db067): Add baseUrl as option
[libs/common/src/services/usernameGeneration.service.ts](https://github.com/bitwarden/clients/compare/master...Djcd:clients:feature/selfhost-anonaddy?expand=1#diff-3cf7fc6c5cd45aaa64dae5837889a2c519948ae03b39c2d5ac4c28cac2ba298d): Add as default URL 'https://app.anonaddy.com' and set it as default on forwarderOption

added by @djsmith85:
- [Add serverurl field to web](https://github.com/bitwarden/clients/pull/4056/commits/01632566e62371a63cd24e9a00b1480c79e3914e)
- [Show anonaddy api errors](https://github.com/bitwarden/clients/pull/4056/commits/138d9bfad31e048c4f02b5e980fddd6150da7cc2)
### Desktop
Screenshot

### Browser
Screenshot 2023-08-16 at 17 33 33

### Web
Screenshot 2023-08-16 at 17 36 16

### CODE REVIEW

**Comment by djsmith85** [comment]
Hi @Djcd thank you very much for your contribution.

I'm sorry this has taken so long. Your changes are looking good so far and we'd like to include your addition.

Would you still be willing to update this PR? Please let me know, so we know if you or one of us should carry this to completion.

The generator.component for web will also need to be updated.

The conflicts should be minor and are mostly because of some recent file moves.

Kind regards,
Daniel

**Comment by Djcd** [comment]
Hi @djsmith85,

sorry for the late answer. Unfortunately i have a lot of work at the moment (building a new house). Could you please continue with the completion? Thank you also in the meantime for taking care!

Kind regards,
David

**Review by aj-rosado — DISMISSED** [review]

**Comment by djsmith85** [comment]
Hi @Djcd,

No problem at all. We've already decided to include this and will carry it to completion. 

Kind regards,
Daniel

**Comment by HeikoBornholdt** [comment]
Would be awesome to see this feature added to Bitwarten soonish

**Review by djsmith85 — DISMISSED** [review]
Reviewed and tested changes made by @Djcd -> Awesome work on this 🚀 
@aj-rosado: I needed to add the server-url field on web and also added a better error message. Please review my changes.

Some weirdness when using a domain , that is not present on the server, will return a HTTP 422 - Unprocessable Content (I used the default against the self-hosted)

**Review by aj-rosado — DISMISSED** [review]
LGTM! 🎉

**Review by djsmith85 — APPROVED** [review]
Some additional feedback/thoughts after passing QA on browser/desktop.

For web (cloud) it doesn't make sense to show the server-url field as the CSP rules will block access to any self-hosted instance of AnonAddy (addy.io)

File: libs/common/src/tools/generator/username/username-generation.service.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ const DefaultOptions: UsernameGeneratorOptions = {
   catchallType: "random",
   forwardedService: "",
   forwardedAnonAddyDomain: "anonaddy.me",
+  forwardedAnonAddyBaseUrl: "https://app.anonaddy.com",
```
Comment (aj-rosado) [inline libs/common/src/tools/generator/username/username-generation.service.ts:27]: I think we should change this to use the new url (https://app.addy.io) like on https://github.com/bitwarden/clients/pull/6027

**Review by aj-rosado — CHANGES_REQUESTED** [review]

File: libs/common/src/tools/generator/username/username-generation.service.ts (hunk 1 above)
Comment (djsmith85) [reply libs/common/src/tools/generator/username/username-generation.service.ts:27]: Nice catch, as #6027 is already in QA, I'm going to wait until that is merged and update the PR here.

**Review by aj-rosado — DISMISSED** [review]
LGTM

**Review by djsmith85 — APPROVED** [review]

**Review by aj-rosado — APPROVED** [review]

**Comment by djsmith85** [comment]
@Djcd Thanks again for your contribution. This has passed QA and is ready to be included in the `2023.10`-release.

**Pull request merged**